### ISSUE TITLE
BuildAbortedException because two binary equal Microsoft.Build.dll differs in LastWriteTime 

### ISSUE THREAD

**Issue by chuchu** [dotnet/msbuild#2771]
### Summary
MSBuild uses a handshake mechanism to determine if two MSBuild nodes can work with each other. This handshake is calculated in the NodeProviderOutOfProc.GetHostHandshake(bool enableNodeReuse) method. One part of the handshake is the LastWriteTime of the Microsoft.MSBuild.dll assembly. This time is read in the Microsoft.Build.Internal.Constants.AssemblyTimestamp property. When you deploy e.g. a custom build tool which brings its own copy of the Microsoft.Build.dll the handshake will never succeed. The reason is that the LastWriteTime depends on the time at which you have installed MSBuild on your system.

When you face this problem a BuildAbortedException is throw with the message “Failed to successfully launch or connect to a child MSBuild.exe process”. In the traces you will see e.g. “Handshake failed. Received 57581269766385075 from host not 66982339790952257. Probably the host is a different MSBuild build”.

### Expected  behavior
From my point of view it would make more sense to just check the full qualified name of the assembly.

### Environment data
Microsoft (R) Build Engine version 15.4.8.50001 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

OS info:
Microsoft Windows [Version 6.1.7601]

**Comment by danmosemsft** [comment]
In your situation is the assembly binary identical, just a duplicate? The handshake is intended to prevent undefined behavior because the contract between nodes is not a stable one, two versions of the binary may have different ideas about it.

**Comment by chuchu** [comment]
My copy of the Microsoft.Build.dll assembly is binary identical with the one which ships with MSBuild. The two files differ only in there last write time which belongs to the file system. This time stamp may change during zip / unzip operation or when you use a copy of another installation.
Unfortunately this time stamp is also included in the handshake. I think this was Ok when the library was installed into the GAC. Then MSBuild and any other executable, which references this assembly will use the same version. But since this assembly is not in the GAC each application need its own copy. I would suggest to use the assembly version instead. Do you really change the contract and keep the version?

**Comment by danmosemsft** [comment]
I don't know whether the file version increments systematically (I'm not on the MSBuild team). How is the custom build tool causing a different version of Microsoft.Build.dll to load? Isn't it being loaded into an (already launched) MSBuild child process, which already has the "real" Microsoft.Build.dll loaded into it?

**Comment by chuchu** [comment]
No, its an own self made console application linked against Microsoft.Build.dll. The build is triggered with BuildManager.DefaultBuildManager.Build(…). This call spawns its own MSBuild processes. Each application (MSBuild and the custom one) loads the  Microsoft.Build.dll from its application directory.

**Comment by fedeazzato** [comment]
I'm experiencing a similar issue.

I've got a custom build tool that I use in my organization, that I've been using since VS2005. Recently I ported the tool to the latest version of msbuild, to be able to build VS2017 projects.

My reference to msbuild dlls is to their respective nuget packages, instead of referencing the dlls from the GAC as in the past. This is because I no longer know where the build tools will be installed on each developer machine.
The assembly version of Microsoft.Build.dll I'm referencing is identical to the one used by the MSBuild process located in "C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin", but they're not binary equal. The one in my build tool is 1,511,592 bytes in size, while the one installed by VS2017 is 1,691,816 bytes.

The weird thing is I'm seeing the BuildAbortedException when BuildManager.Build completes successfully, after the BuildFinished event in the ProjectCollection logger is raised. If I cancel the build with BuildManager.CancelAllSubmissions, the issue doesn't manifest.

Am I doing something wrong?
